File: src/my_project/counterpoint/counterpoint_generator.py
import logging
import random
from collections.abc import Iterator
from dataclasses import dataclass
from functools import cached_property

from my_project.counterpoint import all_measure_validator, validator
from my_project.counterpoint.measure_search.measure_search import MeasureSearch
from my_project.counterpoint.model import (
    AnnotatedMeasure,
    MeasurePosition,
    NoteAnnotation,
    Species,
    ToneType,
)
from my_project.counterpoint.rules import get_measure_rythmn_patterns
from my_project.counterpoint.skeleton_generator import Skeleton, SkeletonGenerator
from my_project.model import (
    Duration,
    FullScore,
    Key,
    Measure,
    Melody,
    Note,
    PartId,
    Pitch,
    Score,
    TimeSignature,
)
from my_project.util import part_range

logger = logging.getLogger(__name__)


@dataclass
class CounterpointGenerator:
    cantus_firmus: list[Pitch]
    cf_part_id: PartId
    key: Key
    species: Species
    part_id: PartId

    measure_search: MeasureSearch
    skeleton_generator: SkeletonGenerator
    rand: random.Random

    _MAX_TRIES = 30

    def __init__(
        self,
        cantus_firmus: list[Pitch],
        cf_part_id: PartId,
        key: Key,
        species: Species,
        part_id: PartId,
        rand: random.Random,
        measure_search: MeasureSearch,
        skeleton_generator: SkeletonGenerator,
    ) -> None:
        self.cantus_firmus = cantus_firmus
        self.cf_part_id = cf_part_id
        self.key = key
        self.species = species
        self.part_id = part_id
        self.rand = rand
        self.measure_search = measure_search
        self.skeleton_generator = skeleton_generator
        pass

    @classmethod
    def default(
        cls,
        cantus_firmus: list[Pitch],
        cf_part_id: PartId,
        key: Key,
        species: Species,
        part_id: PartId,
        seed: int | None = None,
    ) -> "CounterpointGenerator":
        rand = random.Random(seed)
        measure_search = MeasureSearch.default()
        # _measure_length は cantus_firmus の長さから計算されるため、ここで渡す
        _measure_length = len(cantus_firmus)
        skeleton_generator = SkeletonGenerator(
            cantus_firmus=cantus_firmus,
            key=key,
            part_id=part_id,
            rand=rand,
            _measure_length=_measure_length,
        )
        return cls(
            cantus_firmus=cantus_firmus,
            cf_part_id=cf_part_id,
            key=key,
            species=species,
            part_id=part_id,
            rand=rand,
            measure_search=measure_search,
            skeleton_generator=skeleton_generator,
        )

    def __post_init__(self) -> None:
        assert self.cantus_firmus

    @cached_property
    def _measure_length(self) -> int:
        return len(self.cantus_firmus)

    @cached_property
    def _pitch_range(self) -> tuple[Pitch, Pitch]:
        return part_range(self.part_id)

    class AbortAttempt(Exception):
        pass

    def generate_scores(self) -> Iterator[FullScore[NoteAnnotation]]:
        # logger.info("generate_scoresを開始します")

        attempt_count = 0
        while True:
            attempt_count += 1
            # 課題全体の骨格を作成
            skeleton: Skeleton = self.skeleton_generator.generate_skeleton()
            logger.debug(f"Attempt {attempt_count}: Start")

            initial_start_pitch = skeleton.measures[0].notes[0].value
            try:
                yield from self._generate_recursive(skeleton, [], 0, initial_start_pitch)
            except CounterpointGenerator.AbortAttempt:
                logger.debug(f"Attempt {attempt_count}: Aborted. Restarting from scratch.")
                continue

    def _generate_recursive(
        self,
        skeleton: Skeleton,
        completed_measures: list[AnnotatedMeasure],
        measure_index: int,
        current_start_pitch: Pitch,
    ) -> Iterator[FullScore[NoteAnnotation]]:
        log_indents = 2
        indent = " " * (measure_index * log_indents)
        mn_for_log = measure_index + 1

        previous_measure: AnnotatedMeasure | None = None
        previous_cf: Pitch | None = None
        if measure_index > 0:
            previous_measure = completed_measures[measure_index - 1]
            previous_cf = self.cantus_firmus[measure_index - 1]

        current_cf = self.cantus_firmus[measure_index]

        # 最終小節の場合
        if measure_index == self._measure_length - 1:
            last_measure_candidate: AnnotatedMeasure = Melody.of(
                Note(
                    skeleton.measures[measure_index].notes[0].value,
                    Duration.of(4),
                    NoteAnnotation(is_tied_start=False, tone_type=ToneType.HARMONIC_TONE),
                )
            )

            if validator.validate(previous_measure, last_measure_candidate, previous_cf, current_cf):
                logger.debug(f"{indent}Measure {mn_for_log}: [SUCCEED] Last measure created and validated.")
                final_measures = [*completed_measures, last_measure_candidate]
                if not all_measure_validator.validate(final_measures):
                    logger.debug(f"{indent}Attempt failed: [FAILED] All measure validation.]")
                    raise CounterpointGenerator.AbortAttempt()
                else:
                    logger.debug(f"{indent}Attempt succeeded: [SUCCEED] All measures passed validation.]")
                    yield self._to_score(final_measures)
            else:
                logger.debug(f"{indent}Measure {mn_for_log}: [FAILED] Last measure created but failed validation.]")
            return
        # 最終小節以外の場合
        else:
            current_measure_position = self._current_measure_position(measure_index)
            results = self.measure_search.search(
                start_pitch=current_start_pitch,
                start_harmonic_pitch=skeleton.measures[measure_index].notes[0].value,
                next_measure_start_harmonic_pitch=skeleton.measures[measure_index + 1].notes[0].value,
                harmonic_note_names=tuple(skeleton.measures[measure_index].notes[0].attribute),
                key=self.key,
                measure_rythmn_patterns=tuple(get_measure_rythmn_patterns(self.species, current_measure_position)),
                pitch_range=self._pitch_range,
            )
            self.rand.shuffle(results)

            valid_candidates = []
            for chosen_result in results:
                if validator.validate(previous_measure, chosen_result.measure, previous_cf, current_cf):
                    valid_candidates.append(chosen_result)

            if not valid_candidates:
                if len(results) == 0:
                    logger.debug(f"{indent}Measure {mn_for_log}: [FAILED] No candidates found by measure_search.")
                else:
                    logger.debug(
                        f"{indent}Measure {mn_for_log}: [FAILED] Found {len(results)} candidates."
                        " All candidates failed validation."
                    )
                return

            logger.debug(
                f"{indent}Measure {mn_for_log}: [SUCCEED] Found {len(results)} candidates."
                f" Valid {len(valid_candidates)} candidates found."
            )

            for candidate in valid_candidates:
                next_start_pitch = candidate.next_measure_start_pitch
                yield from self._generate_recursive(
                    skeleton,
                    [*completed_measures, candidate.measure],
                    measure_index + 1,
                    next_start_pitch,
                )

    def _current_measure_position(self, current_measure_idx: int) -> MeasurePosition:
        if current_measure_idx == 0:
            return MeasurePosition.FIRST
        elif current_measure_idx == self._measure_length - 1:
            return MeasurePosition.LAST
        elif current_measure_idx == self._measure_length - 2:
            return MeasurePosition.PENULTIMATE
        else:
            return MeasurePosition.MIDDLE

    def _to_score(self, completed_measures: list[AnnotatedMeasure]) -> FullScore[NoteAnnotation]:
        cf_notes: list[Note[Pitch | None, NoteAnnotation]] = [
            Note(
                pitch,
                Duration.of(4),
                NoteAnnotation(is_tied_start=False, tone_type=ToneType.HARMONIC_TONE),
            )
            for pitch in self.cantus_firmus
        ]
        cf_measures: list[AnnotatedMeasure] = [Melody.of(note) for note in cf_notes]

        # body Construction
        time_signature = TimeSignature(2, Duration.of(2))

        parts = {
            self.cf_part_id: [Measure(m) for m in cf_measures],
            self.part_id: [Measure(m) for m in completed_measures],
        }

        body = Score(parts)

        return FullScore(
            key=self.key,
            time_signature=time_signature,
            body=body,
        )

File: backend22/src/main/scala/composer/counterpoint/CounterpointGenerator.scala
package composer.counterpoint

import composer.counterpoint.model._
import composer.counterpoint.search.MeasureSearch
import composer.counterpoint.skeleton.{Skeleton, SkeletonGenerator}
import composer.counterpoint.util.PartUtil
import composer.counterpoint.validation.Validator
import _root_.model.elements._
import _root_.model.containers.{Melody, Note}
import sheet._
import composer.counterpoint.model.Species
import composer.counterpoint.model.MeasureRythmnPattern
import scala.collection.immutable.SeqMap

/** 対位法生成器
  *
  * @param cantusFirmus
  *   定旋律
  * @param cfPart
  *   定旋律のパート
  * @param key
  *   調
  * @param species
  *   対位法の類
  * @param part
  *   対位旋律を生成するパート
  * @param seed
  *   乱数シード
  */
case class CounterpointGenerator(
    cantusFirmus: List[Pitch],
    cfPart: Part,
    key: Key,
    species: Species,
    part: Part,
    seed: Option[Long] = None,
) {

  private val rand = seed.map(s => new scala.util.Random(s)).getOrElse(new scala.util.Random())

  private val measureSearch = MeasureSearch.default
  private val pitchRange    = PartUtil.partRange(part)
  private val measureLength = cantusFirmus.length

  private val MAX_STEPS_PER_ATTEMPT            = 100 // 試行間のバリデーション失敗回数の上限。超えたら最初からやり直し
  private val MAX_VALID_CANDIDATES_PER_MEASURE = 3   // 失敗が連続した際に早く前に戻るために、小節候補を絞る。
  private var currentStepCount                 = 0

  private val skeletonGenerator = new SkeletonGenerator(
    cantusFirmus,
    key,
    pitchRange,
    rand,
    measureLength,
  )

  class AbortAttemptException        extends Exception
  class SucceededAndRestartException extends Exception

  def generateScores: Iterator[SheetMusic] = iterator

  private lazy val iterator: Iterator[SheetMusic] = new Iterator[SheetMusic] {

    var currentBatch: Iterator[SheetMusic] = Iterator.empty

    override def hasNext: Boolean = {
      try {
        if (currentBatch.hasNext) true
        else {
          // Try to generate next batch
          while (!currentBatch.hasNext) {
            currentStepCount = 0
            try {
              // 課題全体の骨格を作成
              val skeleton = skeletonGenerator.generateSkeleton()
              // スケルトンの旋律をロギング (TODO: Logger導入)
              // logger.debug(f"Attempt {attemptCount}: Choose skeleton: ...")
              // logger.debug(f"Attempt {attemptCount}: Start Generate Measures.")

              val initialStartPitch = skeleton.measures.head.elems.head.value.pitch
              currentBatch = generateRecursive(skeleton, Nil, 0, initialStartPitch, None)
              if (currentBatch.hasNext) return true
            } catch {
              case _: AbortAttemptException =>
              // Continue to next attempt
              // logger.debug(f"Attempt {attemptCount}: Aborted. Restarting from scratch.")
              case _: SucceededAndRestartException =>
              // logger.debug(f"Attempt {attemptCount}: Succeeded! Restarting from scratch.")
              // ...
            }
          }
          true
        }
      } catch {
        case _: AbortAttemptException =>
          currentBatch = Iterator.empty
          hasNext
      }
    }

    override def next(): SheetMusic = {
      if (!hasNext) throw new NoSuchElementException("next on empty iterator")
      currentBatch.next()
    }
  }

  private def generateRecursive(
      skeleton: Skeleton,
      completedMeasures: List[AnnotatedMeasure],
      measureIndex: Int,
      currentStartPitch: Pitch,
      previousRythmnPattern: Option[MeasureRythmnPattern],
  ): Iterator[SheetMusic] = {
    currentStepCount += 1
    if (currentStepCount > MAX_STEPS_PER_ATTEMPT) {
      throw new AbortAttemptException()
    }

    val previousMeasure = if (measureIndex > 0) Some(completedMeasures(measureIndex - 1)) else None
    val previousCf      = if (measureIndex > 0) Some(cantusFirmus(measureIndex - 1)) else None
    val currentCf       = cantusFirmus(measureIndex)

    if (measureIndex == measureLength - 1) {
      // 最終小節の場合
      val lastNote: Note[AnnotatedNote] = Note(
        AnnotatedNote(
          Some(skeleton.measures(measureIndex).elems.head.value.pitch),
          NoteAnnotation(isTiedStart = false, toneType = ToneType.HARMONIC_TONE),
        ),
        Duration.of(4),
        Part.Root,
      )
      val lastMeasureCandidate = Melody(List(lastNote))

      if (Validator.validate(previousMeasure, lastMeasureCandidate, previousCf, currentCf)) {
        // logger.debug(f"{indent}Measure {mnForLog}: [SUCCEED] Last measure created and validated.")
        val finalMeasures = completedMeasures :+ lastMeasureCandidate
        if (Validator.validateAll(finalMeasures)) {
          // logger.debug(f"{indent}Attempt succeeded: [SUCCEED] All measures passed validation.]")
          return Iterator(toScore(finalMeasures))
        } else {
          // logger.debug(f"{indent}Attempt failed: [FAILED] All measure validation.]")
          throw new AbortAttemptException()
        }
      } else {
        // logger.debug(f"{indent}Measure {mnForLog}: [FAILED] Last measure created but failed validation.]")
      }
      return Iterator.empty
    } else {
      // 最終小節以外の場合
      val availableRythmnPatterns = getAvailableRythmnPatterns(measureIndex, previousRythmnPattern)

      // Need to extract harmonic note names from skeleton
      // Skeleton measures are Melody[Note[Pitch, ChordWithBass[NoteName]]].
      // We need to extract the chord NoteNames.
      val chordNoteNames = skeleton.measures(measureIndex).elems.head.value.bass.chord.elements

      val startHarmonicPitch            = skeleton.measures(measureIndex).elems.head.value.pitch
      val nextMeasureStartHarmonicPitch = skeleton.measures(measureIndex + 1).elems.head.value.pitch

      val results = measureSearch.search(
        startPitch = currentStartPitch,
        startHarmonicPitch = startHarmonicPitch,
        nextMeasureStartHarmonicPitch = nextMeasureStartHarmonicPitch,
        harmonicNoteNames = chordNoteNames,
        key = key,
        measureRythmnPatterns = availableRythmnPatterns.toSet,
        pitchRange = pitchRange,
      )

      val shuffledResults = rand.shuffle(results)

      val validCandidates = shuffledResults.filter { chosenResult =>
        Validator.validate(previousMeasure, chosenResult.measure, previousCf, currentCf)
      }

      if (validCandidates.isEmpty) return Iterator.empty

      val topCandidates = validCandidates.take(MAX_VALID_CANDIDATES_PER_MEASURE)

      topCandidates.iterator.flatMap { candidate =>
        generateRecursive(
          skeleton,
          completedMeasures :+ candidate.measure,
          measureIndex + 1,
          candidate.nextMeasureStartPitch,
          Some(candidate.rythmnPattern),
        )
      }
    }
  }

  private def getAvailableRythmnPatterns(
      measureIndex: Int,
      previousRythmnPattern: Option[MeasureRythmnPattern],
  ): List[MeasureRythmnPattern] = {
    // We need to implement Rules.getMeasureRythmnPatterns logic.
    // I haven't implemented Rules.scala yet.
    // I can implement the logic here or in a separate object.
    // Given the imports, I should probably implement the logic here for simplicity or create Rules object.
    // Let's create a private helper for now, mimicking Rules.get_measure_rythmn_patterns.

    val currentMeasurePosition = this.currentMeasurePosition(measureIndex)
    var availablePatterns      = getPatternsForSpecies(species, currentMeasurePosition)

    if (previousRythmnPattern.isDefined) {
      val prev = previousRythmnPattern.get
      availablePatterns = availablePatterns.filterNot(_ == prev)

      val isPrevTiedToNext = prev.measureRythmn.isNextTied
      availablePatterns = availablePatterns.filter(p => p.measureRythmn.isPreviousTied == isPrevTiedToNext)
    }
    availablePatterns
  }

  private def currentMeasurePosition(currentMeasureIdx: Int): MeasurePosition = {
    if (currentMeasureIdx == 0) MeasurePosition.FIRST
    else if (currentMeasureIdx == measureLength - 1) MeasurePosition.LAST
    else if (currentMeasureIdx == measureLength - 2) MeasurePosition.PENULTIMATE
    else MeasurePosition.MIDDLE
  }

  private def getPatternsForSpecies(species: Species, position: MeasurePosition): List[MeasureRythmnPattern] = {
    // Port logic from rules.py
    species match {
      case Species.FIRST_SPECIES  => List(MeasureRythmnPattern.R_1)
      case Species.SECOND_SPECIES =>
        position match {
          case MeasurePosition.FIRST => List(MeasureRythmnPattern.R_rr2)
          case MeasurePosition.LAST  => List(MeasureRythmnPattern.R_1)
          case _                     => List(MeasureRythmnPattern.R_22)
        }
      case Species.THIRD_SPECIES =>
        position match {
          case MeasurePosition.FIRST => List(MeasureRythmnPattern.R_r444)
          case MeasurePosition.LAST  => List(MeasureRythmnPattern.R_1)
          case _                     => List(MeasureRythmnPattern.R_4444)
        }
      case Species.FOURTH_SPECIES =>
        position match {
          case MeasurePosition.FIRST  => List(MeasureRythmnPattern.R_rr2, MeasureRythmnPattern.R_rr2t)
          case MeasurePosition.MIDDLE =>
            List(
              MeasureRythmnPattern.R_22,
              MeasureRythmnPattern.R_t22,
              MeasureRythmnPattern.R_22t,
              MeasureRythmnPattern.R_t22t,
            )
          case MeasurePosition.PENULTIMATE => List(MeasureRythmnPattern.R_22, MeasureRythmnPattern.R_t22)
          case MeasurePosition.LAST        => List(MeasureRythmnPattern.R_1)
        }
      case Species.FIFTH_SPECIES =>
        position match {
          case MeasurePosition.FIRST =>
            List(
              MeasureRythmnPattern.R_rr2,
              MeasureRythmnPattern.R_rr2t,
              MeasureRythmnPattern.R_r444,
              MeasureRythmnPattern.R_r42,
            )
          case MeasurePosition.MIDDLE =>
            List(
              MeasureRythmnPattern.R_22,
              MeasureRythmnPattern.R_t22,
              MeasureRythmnPattern.R_22t,
              MeasureRythmnPattern.R_t22t,
              MeasureRythmnPattern.R_4444,
              MeasureRythmnPattern.R_244,
              MeasureRythmnPattern.R_442,
              MeasureRythmnPattern.R_t4444,
              MeasureRythmnPattern.R_t244,
              MeasureRythmnPattern.R_t442,
              MeasureRythmnPattern.R_4444t,
              MeasureRythmnPattern.R_244t,
              MeasureRythmnPattern.R_2488,
              MeasureRythmnPattern.R_4882,
              MeasureRythmnPattern.R_t2488,
              MeasureRythmnPattern.R_t4882,
              MeasureRythmnPattern.R_4882t,
              MeasureRythmnPattern.R_2d4,
              MeasureRythmnPattern.R_2d88,
            )
          case MeasurePosition.PENULTIMATE =>
            List(
              MeasureRythmnPattern.R_22,
              MeasureRythmnPattern.R_t22,
              MeasureRythmnPattern.R_4444,
              MeasureRythmnPattern.R_244,
              MeasureRythmnPattern.R_442,
              MeasureRythmnPattern.R_t4444,
              MeasureRythmnPattern.R_t244,
              MeasureRythmnPattern.R_t442,
              MeasureRythmnPattern.R_2488,
              MeasureRythmnPattern.R_4882,
              MeasureRythmnPattern.R_t2488,
              MeasureRythmnPattern.R_t4882,
              MeasureRythmnPattern.R_2d4,
              MeasureRythmnPattern.R_2d88,
            )
          case MeasurePosition.LAST => List(MeasureRythmnPattern.R_1)
        }
    }
  }

  private def toScore(completedMeasures: List[AnnotatedMeasure]): SheetMusic = {
    // Convert AnnotatedMeasure (Melody[AnnotatedNote, Unit, Note[AnnotatedNote, Unit]]) to sheet.Measure
    // AttributedValue(value: Pitch | Rest, attr: Option[ScoreAttrs])

    def convertMeasure(annotated: AnnotatedMeasure): sheet.Measure = {
      val newElems = annotated.elems.map { note =>
        val value: Pitch | Rest = note.value.value.getOrElse(Rest)
        val attrs               = ScoreAttrs(
          isTiedStart = note.value.annotation.isTiedStart,
          graces = Nil, // Grace notes not supported yet
        )
        Note(AttributedValue(value, Some(attrs)), note.duration, note.part)
      }
      sheet.Measure(Melody(newElems))
    }

    val partMeasures = completedMeasures.map(convertMeasure)

    val cfMeasures = cantusFirmus.map { pitch =>
      val note = Note(
        AttributedValue(pitch, Some(ScoreAttrs(false))),
        Duration.of(4),
        cfPart,
      )
      sheet.Measure(Melody(List(note)))
    }

    // 2/2 time signature as per Python code
    val timeSig = sheet.TimeSignature(2, Duration.of(2))

    val parts = Seq(
      cfPart -> cfMeasures,
      part   -> partMeasures,
    )

    val partOrdering = Part.ordering("Soprano", "Alto", "Tenor", "Bass")

    SheetMusic(
      key = key,
      timeSignature = timeSig,
      timeSignatureEvents = Nil,
      keySignatureEvents = Nil,
      body = PartMapScore(SeqMap.from(parts.sortBy(_._1)(using partOrdering))),
      title = None,
    )
  }
}
